### PR TITLE
feat: utilization/memory usage on CPU and GPU

### PR DESCRIPTION
**Pull request by GustavBaumgart**
## Description

The metric collector now runs two additional threads to track GPU/CPU utilization and memory usage. So far, GPU utilization is still not specific to the process that is running.

Additionally, the metric collector was refactored in order to store samples in a variable stat_log. Names of metrics now use a period to separate the metric type and alias.

setup.py has been updated to include gpustat and psutil.

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: lib/python/flame/monitor/metric_collector.py [hunk 1]
```diff
@@ -16,31 +16,123 @@
 """Metric Collector."""
 
 import logging
+import os
+import time
+import threading
+import psutil
+from collections import defaultdict
+from flame.common.util import MLFramework, get_ml_framework_in_use
+from gpustat.nvml import pynvml as N
```
Comment (myungjin) [inline lib/python/flame/monitor/metric_collector.py:25]: Looks like gpustat's module is not used. If pynvml is the only module used, why don't we install pynvml package directly instead of gpustat?

File: lib/python/setup.py [hunk 2]
```diff
@@ -18,38 +18,40 @@
...
+        "aiostream",
+        "boto3",
+        "cloudpickle",
+        "diskcache",
+        "mlflow==2.0.1",
+        "paho-mqtt",
+        "protobuf==3.19.5",
+        "grpcio==1.51.1",
+        "pydantic",
+        "gpustat",
```
Comment (myungjin) [inline lib/python/setup.py:44]: please consider to use pynvml instead of gpustat if gpustat's functionality is not used.

File: lib/python/flame/monitor/metric_collector.py [hunk 3]
```diff
@@ -16,31 +16,123 @@
...
+
+        # GPU monitoring
+        gpu_thread = threading.Thread(target=self.gather_gpu_stats)
+        ml_framework = get_ml_framework_in_use()
+        if ml_framework == MLFramework.PYTORCH:
+            import torch
+
+            if torch.cuda.is_available():
+                gpu_thread.start()
+        elif ml_framework == MLFramework.TENSORFLOW:
```
Comment (myungjin) [inline lib/python/flame/monitor/metric_collector.py:48]: Is it necessary to check what ml framework is in use for gpu stat monitoring?

I think it may be enough to revise lines 56 and 57.

At N.nvmlInit(), you can check whether htere is an error (e.g., pynvml.nvml.NVMLError_LibraryNotFound).

At dcount = N.nvmlDeviceGetCount(), you can check whether dcount is zero or not.

Then, you may not need to know which ml framework is in use.

File: lib/python/flame/monitor/metric_collector.py [hunk 4]
```diff
@@ -16,31 +16,123 @@
 """Metric Collector."""
 
 import logging
+import os
+import time
+import threading
+import psutil
+from collections import defaultdict
+from flame.common.util import MLFramework, get_ml_framework_in_use
```
Comment (myungjin) [inline lib/python/flame/monitor/metric_collector.py:24]: It's a good idea to separate standard packages from others (3rd party ones and our own ones).
Perhaps if you apply `black`, it may do reformatting.

**Review by myungjin — COMMENTED** [review]
left a few comments. Please consider to address them.

File: lib/python/flame/monitor/metric_collector.py [hunk 5]
```diff
@@ -16,31 +16,123 @@
...
         logger.debug(f"Saving state_dict[{key}] = {value}")
-    
+
     def accumulate(self, mtype, alias, value):
         key = self.get_key(mtype, alias)
         self.state_dict[key] = value + self.state_dict.get(key, 0)
         logger.debug(f"Accumulating metric state_dict[{key}] = {self.state_dict[key]}")
-    
+
+    def save_log_statistics(self):
```
Comment (jaemin-shin) [inline lib/python/flame/monitor/metric_collector.py:119]: can we add more than this? like quartile (25, 50, 75), median, std ? Using numpy array makes lot more easier, as it has its own functions like np.mean, np.min, np.median, np.quartile etc.

**Review by jaemin-shin — COMMENTED** [review]
Left some comments that could be considered, but generally LGTM.

File: lib/python/flame/monitor/metric_collector.py (hunk 1 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:25]: I don't know why, but it looks like because of the way it is packaged in gpustat, gpustat is a lot better at avoiding errors that usually come up when I try to use pynvml directly. I think it tries to use several methods from pynvml but then has backup plans as well.

File: lib/python/setup.py (hunk 2 above)
Comment (GustavBaumgart) [reply lib/python/setup.py:44]: Will probably not change this for now.

File: lib/python/flame/monitor/metric_collector.py (hunk 4 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:24]: I believe I applied `black`, does it usually do that for you?

File: lib/python/flame/monitor/metric_collector.py [hunk 6]
```diff
@@ -16,31 +16,123 @@
...
+
+            if torch.cuda.is_available():
+                gpu_thread.start()
+        elif ml_framework == MLFramework.TENSORFLOW:
+            import tensorflow
+
+            if tensorflow.config.experimental.list_physical_devices("GPU"):
+                gpu_thread.start()
+
+    def gather_gpu_stats(self, interval=1):
```
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:54]: This, I am not certain of. So far, I haven't run any large-scale example along with the metric collection added in. Maybe I should do that before.

File: lib/python/flame/monitor/metric_collector.py [hunk 7]
```diff
@@ -16,31 +16,123 @@
...
+        dcount = N.nvmlDeviceGetCount()
+
+        while True:
+            for d in range(dcount):
+                handle = N.nvmlDeviceGetHandleByIndex(d)
+
+                # GPU utilization
+                # TO DO: implement metric gathering for process-specific utilization of the GPUs
+                try:
+                    self.stat_log[f"gpu{d}_utilization"].append(
```
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:66]: This would be some additional overhead cost, but I can try.

File: lib/python/flame/monitor/metric_collector.py (hunk 5 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:119]: Yeah this was just an initial sort of idea for the stat collection. I can change it to numpy if it's in the requirements.

File: lib/python/flame/monitor/metric_collector.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:66]: Actually, do you think it would be misleading given the way I measure memory usage of the process?

[the remark this answers is not in the document]
Comment (jaemin-shin) [reply lib/python/flame/monitor/metric_collector.py:66]: Memory usage looks fine. My point is that we may need to subtract the utilization from running GPUs that aren't running any Flame processes, which we could at least know. We could do this by post-processing by looking at 0 memory usage, but I would prefer doing this within a same loop, as checking pid wouldn't take much of a overhead.

File: lib/python/flame/monitor/metric_collector.py (hunk 3 above)
Comment (GustavBaumgart) [reply lib/python/flame/monitor/metric_collector.py:48]: Update: will check overhead and change this

File: lib/python/flame/monitor/metric_collector.py [hunk 8]
```diff
@@ -16,31 +16,131 @@
...
+                except:
+                    logger.debug(f"failed to get GPU utilization of GPU {d}")
+
+                # GPU memory usage of process
+                try:
+                    self.stat_log[f"gpu{d}_memory"].append(
+                        sum(
+                            [
+                                proc.usedGpuMemory
+                                for proc in N.nvmlDeviceGetComputeRunningProcesses(
```
Comment (myungjin) [inline lib/python/flame/monitor/metric_collector.py:85]: `nvmlDeviceGetComputeRunningProcesses` is called twice (one here and the other in line 72).
I think it can be called only once (e.g., in line 63) and compute both memory and utilization at the same time.
Also, this may be pythonic way, but the code structure look so complicated: embedding for and if statements as an argument of a function (e.g., append) with `sum` function as well.

Can you simplify the code?

**Review by myungjin — COMMENTED** [review]
left one comment

**Review by myungjin — COMMENTED** [review]
left one comment

**Review by myungjin — APPROVED** [review]
lgtm

**Review by myungjin — APPROVED** [review]
lgtm
